### PR TITLE
Reduce heap allocations and use modern cpp

### PR DESCRIPTION
**Pull request by ronag**

### CODE REVIEW

**Comment by vweevers** [comment]
It'd be good to separate pure cosmetic changes (that can potentially be merged quickly) from ones that benefit performance, because the latter require benchmarks.

**Comment by ronag** [comment]
Do you have any benchmarking script?

**Comment by vweevers** [comment]
Yes, but it comes with instructions - that I'll write later

**Comment by ronag** [comment]
TBH I don't really see anything here that would cause perf regression.

**Comment by ralphtheninja** [comment]
I think it looks good but there are a lot of changes for a single commit.

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -8,7 +8,7 @@ jobs:
       matrix:
         # At the time of writing macos-latest is mac 10; we need 11 to build a universal binary.
         os: [ubuntu-latest, macos-11, windows-latest]
-        node: [12, 14, 16]
+        node: [14, 16]
```
Comment (ronag) [inline .github/workflows/test.yml:11]: For some reason doesn't build on v12? Old cpp version?

**Comment by ronag** [comment]
If you want more step by step then https://github.com/Level/classic-level/pull/20 and https://github.com/Level/classic-level/pull/24 is some steps towards this.

**Comment by ronag** [comment]
I'l lsplit this up

**Pull request closed without merging**